src/Kestrel.Core/KestrelServer.cs

```diff
@@ -135,15 +135,17 @@ private static ServiceContext CreateServiceContext(IOptions<KestrelServerOptions
 
                 async Task OnBind(ListenOptions endpoint)
                 {
-                    // Add the connection limit middleware
-                    endpoint.UseConnectionLimit(ServiceContext);
-
-                    // Configure the user delegate
-                    endpoint.Configure(endpoint);
-
                     // Add the HTTP middleware as the terminal connection middleware
                     endpoint.UseHttpServer(endpoint.ConnectionAdapters, ServiceContext, application);
 
+                    var connectionDelegate = endpoint.Build();
+
+                    // Add the connection limit middleware
+                    if (Options.Limits.MaxConcurrentConnections.HasValue)
+                    {
+                        connectionDelegate = new ConnectionLimitMiddleware(connectionDelegate, Trace, Options.Limits.MaxConcurrentConnections.Value).OnConnectionAsync;
+                    }
```

Comment:
@davidfowl The rest of this PR will likely need to be undone because upgraded connections need to be removed from the normal connection count, but the change here to remove the call to `endpoint.Configure` I think is solid.